driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -807,18 +652,233 @@ void onBulkWriteCommandErrorWithoutResponse(final MongoException exception) {
+                    int finalModelIndexInBatch = modelIndexInBatch;
+                    writeResult = writersProviderAndLimitsChecker.tryWrite((opsWriter, nsInfoWriter) -> {
+                        batchEncoder.encodeWriteModel(opsWriter, namespacedModel.getModel(), finalModelIndexInBatch, namespaceIndexInBatch);
+                        if (writeNewNamespace) {
+                            nsInfoWriter.writeStartDocument();
+                            nsInfoWriter.writeString("ns", namespace.getFullName());
+                            nsInfoWriter.writeEndDocument();
+                        }
+                        return finalModelIndexInBatch + 1;
+                    });
+                    if (writeResult == FAIL_LIMIT_EXCEEDED) {
+                        batchEncoder.reset(finalModelIndexInBatch);
+                    } else {
+                        maxModelIndexInBatch = finalModelIndexInBatch;
+                        if (commandIsRetryable && doesNotSupportRetries(namespacedModel)) {
+                            commandIsRetryable = false;
+                            logWriteModelDoesNotSupportRetries();
+                        }
+                    }
+                }
+                return new EncodeDocumentsResult(
+                        // we will execute more batches, so we must request a response to maintain the order of individual write operations
+                        options.isOrdered() && maxModelIndexInBatch < models.size() - 1,
+                        commandIsRetryable
+                                ? singletonList(new BsonElement("txnNumber", new BsonInt64(doIfCommandIsRetryableAndAdvanceGetTxnNumber.get())))
+                                : emptyList());
+            }
+
+            private static boolean doesNotSupportRetries(final AbstractClientNamespacedWriteModel model) {
+                return model instanceof ConcreteClientNamespacedUpdateManyModel || model instanceof ConcreteClientNamespacedDeleteManyModel;
+            }
+
+            /**
+             * The server supports only the {@code update} individual write operation in the {@code ops} array field, while the driver supports
+             * {@link ClientNamespacedUpdateOneModel}, {@link ClientNamespacedUpdateOneModel}, {@link ClientNamespacedReplaceOneModel}.
+             * The difference between updating and replacing is only in the document specified via the {@code updateMods} field:
+             * <ul>
+             *     <li>if the name of the first field starts with {@code '$'}, then the document is interpreted as specifying update operators;</li>
+             *     <li>if the name of the first field does not start with {@code '$'}, then the document is interpreted as a replacement.</li>
+             * </ul>
+             */
+            private static final class OpsFieldNameValidator implements FieldNameValidator {
```

Comment:
As discussed on the call, let's add a link to the [spec requirement](https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md#update-vs-replace-document-validation) for clarity.

Reply:
Done in 38c880d988c2c7f9d4a8716b4155ef36f43de77b.